src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes implementation for the NAT64 translator.
+ *
+ */
+
+#include "nat64_translator.hpp"
+
+#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
+
+#include <openthread/border_router.h>
+#include <openthread/logging.h>
+
+#include "common/code_utils.hpp"
+#include "common/locator_getters.hpp"
+#include "common/uptime.hpp"
+#include "net/checksum.hpp"
+#include "net/ip4_types.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+namespace BorderRouter {
+
+RegisterLogModule("Nat64");
+
+Nat64Translator::Nat64Translator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mAvailableAddressCount(0)
+{
+    mNat64Prefix.Clear();
+    mIp4Cidr.Clear();
+}
+
+Nat64Translator::Result Nat64Translator::HandleOutgoing(Message &aMessage)
+{
+    Error           err = kErrorNone;
+    Result          res = Result::kDrop;
+    Ip6::Header     ip6Header;
+    Ip4::Header     ip4Header;
+    AddressMapping *mapping = nullptr;
+
+    // ParseForm will do basic checks for the message, including the packet length and IP protocol version.
+    err = ip6Header.ParseFrom(aMessage);
+    if (err != kErrorNone)
+    {
+        LogWarn("outgoing packet is not a valid IPv6 packet, drop");
+        ExitNow(res = Result::kDrop);
+    }
+
+    if (!mNat64Prefix.IsValidNat64() || !ip6Header.GetDestination().MatchesPrefix(mNat64Prefix))
+    {
+        ExitNow(res = Result::kForward);
+    }
+
+    if (mIp4Cidr.mLength == 0)
+    {
+        // The NAT64 translation is bypassed (will be handled externally)
+        LogWarn("no IPv4 CIDR for NAT64 configured, forward to upper layer");
+        ExitNow(res = Result::kForward);
+    }
+
+    if (ip6Header.GetHopLimit() <= 1)
+    {
+        LogDebg("outgoing packet hop limit reached, drop");
+        ExitNow(res = Result::kDrop);
+    }
+    ip6Header.SetHopLimit(ip6Header.GetHopLimit() - 1);
+
+    mapping = GetMapping(ip6Header.GetSource(), /* aTryCreate */ true);
+    if (mapping == nullptr)
+    {
+        LogWarn("failed to get a mapping for %s (mapping pool full?)", ip6Header.GetSource().ToString().AsCString());
+        ExitNow(res = Result::kDrop);
+    }
+
+    aMessage.RemoveHeader(sizeof(Ip6::Header));
+
+    ip4Header.Clear();
+    ip4Header.InitVersionIhl();
+    ip4Header.GetSource().SetBytes(mapping->mIp4.GetBytes());
```

Comment:
I think this works as well and is simpler to read:
```suggestion
    ip4Header.SetSource(mapping->mIp4);
```